apps/src/weblab/WebLab.js

```diff
@@ -0,0 +1,213 @@
+  this.studioApp_ = studioApp;
+  this.studioApp_.reset = this.reset.bind(this);
+};
+
+/**
+ * Returns the global singleton. For use by Bramble host to get the Web Lab interface.
+ * Set on the window object so that Bramble host can get to it from inside its iframe.
+ */
+window.getWebLab = function () {
+  return webLab_;
+};
+
+/**
+ * Initialize this WebLab instance.  Called on page load.
+ */
+WebLab.prototype.init = function (config) {
+  if (!this.studioApp_) {
+    throw new Error("WebLab requires a StudioApp");
+  }
+
+  this.skin = config.skin;
+  this.level = config.level;
+
+  this.brambleHost = null;
+
+  if (this.level.lastAttempt) {
+    this.startSources = this.level.lastAttempt;
+  } else if (this.level.startSources && this.level.startSources.length > 0) {
+    try {
+      this.startSources = JSON.parse(this.level.startSources);
+    } catch (err) {
+      console.error("Unable to parse startSources list", err);
+    }
+  }
+
+  config.usesAssets = true;
+
+  config.makeYourOwn = false;
+  config.centerEmbedded = false;
+  config.wireframeShare = true;
+  config.noHowItWorks = true;
+
+  config.getCodeAsync = this.getCodeAsync.bind(this);
+
+  // Provide a way for us to have top pane instructions disabled by default, but
+  // able to turn them on.
+  config.showInstructionsInTopPane = true;
+  config.noInstructionsWhenCollapsed = true;
+
+  const onMount = () => {
+    this.setupReduxSubscribers(this.studioApp_.reduxStore);
+
+    // TODO: understand if we need to call studioApp
+    // Other apps call studioApp.init(). That sets up UI that is not present Web Lab (run, show code, etc) and blows up
+    // if we call it. It's not clear there's anything in there we need, although we may discover there is and need to refactor it
+    // this.studioApp_.init(config);
+  };
+
+  // Push initial level properties into the Redux store
+  this.studioApp_.setPageConstants(config, {
+    channelId: config.channel,
+    isProjectLevel: !!config.level.isProjectLevel,
+  });
+
+  function onUndo() {
+    this.brambleHost.undo();
+  }
+
+  function onRedo() {
+    this.brambleHost.redo();
+  }
+
+  function onShowPreview() {
+    this.brambleHost.hideTutorial();
+    // temporarily, register a "change" when the preview or tutorial buttons are pressed. TODO: hook up onProjectChanged to Bramble
+    this.onProjectChanged();
+  }
+
+  function onShowTutorial() {
+    this.brambleHost.showTutorial();
+    // temporarily, register a "change" when the preview or tutorial buttons are pressed. TODO: hook up onProjectChanged to Bramble
+    this.onProjectChanged();
+  }
+
+  let inspectorOn = false;
+  function onToggleInspector() {
+    inspectorOn = !inspectorOn;
+    if (inspectorOn) {
+      this.brambleHost.enableInspector();
+    } else {
+      this.brambleHost.disableInspector();
+    }
+  }
+
+  ReactDOM.render((
+    <Provider store={this.studioApp_.reduxStore}>
+      <WebLabView
+        onUndo={onUndo.bind(this)}
```

Comment:
could this just be `this.brambleHost.undo.bind(this.brambleHost)` instead of creating a new local function that just calls brambleHost.undo? Same with redo.